yoyodyne/models/lstm.py

```diff
@@ -402,7 +417,11 @@ def forward(self, batch: batches.PaddedBatch) -> torch.Tensor:
             )
         else:
             predictions = self.decode(
-                len(batch), batch.source.mask, encoder_out, batch.target.padded
+                len(batch),
+                batch.source.mask,
+                encoder_out,
+                batch.target.padded,
```

Comment:
Suggest making this:

`batch.target.padded if teacher_forcing else None`

and the removing the following argument, and adjusting interface to respect  that.

Reply:
Sorry I am a bit all over the place today, but also just realized that a user *should* be able to specify `--target_col=0` to control this from the CLI. Idk that we ever actually tested that though.

Reply:
Yes, that'd be great if we could just document that and have it work that way! (I'd also log the teacher/student status in the `train` or `predict` main.)

Reply:
On reflection I don't think this will (or could) work for training, just prediction. In training, we need to read in the target whether or not we're using it during decoding. In prediction though, this might...

Reply:
Right agreed. Maybe I need to think this through better from the perspective of what is most logical for a user.

I think the teacher forcing flag is useful for making it easy to understand the code during training (check the validation step code, see that we set teacher forcing to False during validation) -- and that this is a pretty natural behavior.

During prediction, we _could_ require a user to set the target col to 0, but then again, I think it is also a natural assumption that `predict_step` should not have access to the targets, even if they are in the file. Maybe the best course of action is to leave it up to the user, but log a warning if target is not zero'd out. 

My only concern is if this warning will be understandable :D. I think the default behavior would be that we predict with teacher_forcing=False, but that, as in validation, we can stop early once we have predicted at least as many tokens as the target length.

Reply:
At the highest level I think the interface should be:

* `--forcing` defaults to "default" (teacher during training, student during prediction)
* `--forcing student` does what you think regardless of whether we're doing training or prediction
* `--forcing teacher` ... ditto

Regardless of how the dataset args are passed, you should need a valid target column for teacher forcing and it shouldn't matter for student forcing.

Reply:
Been trying to add this as an argument that we call from train/predict, but I am unsure of how pass an argument through the training loop in lightning, and cannot find anything in docs/issues.

Now I am wondering if we should consider `teacher_forcing` as only valid during training, and add it as a model attribute (if we later extend it to a teacher forcing "ratio", that can still live on the model). Then, the question still remains, where should we tell the forward/decode method to ignore the gold symbols if they exist?

One clean solution (1) is to always force `teacher_forcing=False` in the validate/predict steps like we are doing now. Instead, I suppose we could (2) check if the model is in `eval()` mode? I think (1) makes the most sense. But I keep worrying if the semantics of this is confusing. 

Additional point: I would rather keep the argument as `--teacher_forcing` and have it as either a bool or float if we want a ratio. I think having forcing types is a bit weird since IDK if `student_forcing` is a common term, not if there could be any other possible type of forcing that is not expressed by an "amount" of teacher forcing.

Reply:
Jumping in on this, unless it was changed a few prs ago, I believe I wrote into `predict` that `target-col` is automatically set to `0` just for these concerns. So during the predict stage, the upstream should be stopping any presence of target.

Reply:
I think what we have here is good for now...resolving.